### PR TITLE
MNT: _plot_glyphs slow on Windows

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
On Windows, plotting coreg's glyphs with all options (`orient_glyphs`, `scale_by_distance` and `mark_inside`) takes an abnormally long time.
This PR tries to address this.

It's an item of #8833

### CODE REVIEW

File: mne/gui/_coreg.py [hunk 1]
```diff
@@ -903,6 +905,7 @@ def _update_projection_surface(self):
             tris=self.coreg._bem_low_res["tris"],
             nn=self.coreg._bem_low_res["nn"]
         )
+        self._check_inside = _CheckInside(self._head_geo)
```
Comment (larsoner) [inline mne/gui/_coreg.py:908]: Good idea! Instantiating this once should make it a *lot* faster. The only caveat is that you probably need to change it when the `_head_geo` changes (dense or not).

At one point I think we maybe *always* used the non-dense head for these computations. It definitely makes things faster, but it does mess up things like the tip of the nose. So maybe we should just change it with the head surface changes...

Comment (GuillaumeFavelier) [reply mne/gui/_coreg.py:908]: > Instantiating this once should make it a lot faster.

That's what I thought at first but then I realized maybe most of the computation is not done in `_CheckInside.__init__()` but in `_CheckInside.__call__()`? I'll keep digging.

**Comment by larsoner** [comment]
Happy to let you test :) My first API idea is to add a `resize_test=0.` param to the `__init__`, where if > 0, it uses the given `resize_test` value as the shrink-inward and expand-outward displacement (e.g., 0.001 if units are meters or 1 if units are mm for the case of the head surface) of all points toward the center of mass of the surface. This COM is already calculated elsewhere for the sphere test, so it shouldn't be too hard to shrink and expand the surfaces, set up the nearest-neighbor classes, and then use them for a test...

**Comment by GuillaumeFavelier** [comment]
Considering the average of 10 runs, redrawing {'eeg', 'hsp', 'hpi', 'head_fids', 'helmet', 'head', 'mri_fids'}(1):

x | main | PR
--|--|--
linux | 2.36s | 1.63
windows | - | -

*(1) not everything in the list is actually drawn, it depends on the default flags*

**Comment by GuillaumeFavelier** [comment]
It's a step forward but the big difference between linux and windows needs to be investigated further.

**Comment by larsoner** [comment]
I have one simple idea using octants for how to speed it up. I'll open a PR shortly hopefully...

**Comment by agramfort** [comment]
I would look into VTK here. Eg

https://stackoverflow.com/questions/52699495/check-if-a-point-is-inside-and-object-using-vtk

**Comment by GuillaumeFavelier** [comment]
> That's what I thought at first but then I realized maybe most of the computation is not done in _CheckInside.__init__() but in _CheckInside.__call__()?

I'm so sorry @larsoner, my assumption was wrong.

**Comment by larsoner** [comment]
> So __init__() is way slower than __call__() (115s against 9s) and the line:

This is great actually! The `_CheckInside` only needs to change (in principle) when grow_hair or MRI scaling changes. So I would make this the case in coreg, probably by 1.

1. making a `self._check_inside` a new property (trait?) that depends on `grow_hair` and `scale` params
2. Pass this _CheckInside instance directly to the 3D glyph plotting function (which would now accept not just a `surf`, as it does now, but also a `_CheckInside` instance to use `__call__` directly without instantiating)

WDYT @GuillaumeFavelier ? Some variant of this should help a lot, right?

**Comment by GuillaumeFavelier** [comment]
I think that `2.` is already implemented but `1.` should help definitely :+1:

**Comment by larsoner** [comment]
What we should probably really use is

https://vtk.org/doc/nightly/html/classvtkSelectEnclosedPoints.html

But we need to make sure our mesh is properly closed using `IsSurfaceClosed()` (we should probably already be doing this sort of thing anyway), and fix it if it's not.

@GuillaumeFavelier can you check to see if this works in place of our delaunay+_points_outside_surface? (You can leave the sphere path as is, it's plenty fast and pretty simple/harmless).

**Comment by GuillaumeFavelier** [comment]
> can you check to see if this works

Yes, I'll try it next

**Comment by larsoner** [comment]
See also:

https://public.kitware.com/pipermail/vtkusers/2010-April/059336.html
https://vtk.org/doc/nightly/html/classvtkFillHolesFilter.html

**Comment by larsoner** [comment]
@GuillaumeFavelier okay for me to port over my test updates and one small fix from https://github.com/mne-tools/mne-python/pull/10422/files over here?

**Comment by GuillaumeFavelier** [comment]
No problem @larsoner

**Comment by larsoner** [comment]
Okay, pushed a commit with my changes. It looks like something is still wrong with the _check_inside caching/update, it should not be called during ICP at all. When I do this command then click "lock fiducials" and then "Fit ICP", I get a new setup call with each ICP iteration:
```
Setting up interior check for 10242 points...
Setup complete in 247.7 ms
Setting up interior check for 2562 points...
Setup complete in 50.4 ms
Setting up interior check for 2562 points...
Setup complete in 49.5 ms
Checking surface interior status for 78 points...
    Found  1/78 point  that fit inside a sphere of radius   79.1 mm
    Found 61/77 points outside using surface Qhull
    Found  2/16 points outside using solid angles
    Total 15/78 points inside the surface
Interior check completed in 41.2 ms
Setting up interior check for 2562 points...
Setup complete in 49.0 ms
Checking surface interior status for 78 points...
    Found  1/78 point  that fit inside a sphere of radius   79.1 mm
    Found 55/77 points outside using surface Qhull
    Found  3/22 points outside using solid angles
    Total 20/78 points inside the surface
Interior check completed in 26.3 ms
Setting up interior check for 2562 points...
Setup complete in 50.3 ms
Checking surface interior status for 78 points...
    Found  1/78 point  that fit inside a sphere of radius   79.1 mm
    Found 53/77 points outside using surface Qhull
    Found  3/24 points outside using solid angles
    Total 22/78 points inside the surface
Interior check completed in 27.9 ms
...
```
@GuillaumeFavelier do you want to look and try to fix this?

In any case, the `verbose=True`s in here (which should be reverted before merge) should allow us to test the VTK Delaunay + interior point check stuff more easily. @GuillaumeFavelier are you also looking into that, or should I?

File: mne/viz/_3d.py [hunk 2]
```diff
@@ -1056,19 +1059,22 @@ def _orient_glyphs(pts, surf, project_to_surface=False, mark_inside=False):
...
         scalars, vectors, proj_pts = _orient_glyphs(loc, surf, project_points,
-                                                    mark_inside)
+                                                    mark_inside, check_inside,
+                                                    nearest)
         if mark_inside:
             from matplotlib.colors import ListedColormap
             color = np.append(color, 1)
             colormap = ListedColormap(np.array([(0, 0, 0, 1,), color]))
             color = None
+            clim = [0, 1]
```
[the remark this answers is not in the document]
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py:1073]: Okay, good catch.

**Comment by GuillaumeFavelier** [comment]
> It looks like something is still wrong with the _check_inside caching/update, it should not be called during ICP at all.

It's because `2.` is not implemented yet. `_CheckInside.__init__()` is called at each call of `_add_head_surface()` (in `_update_surface_projection()`.

**Comment by larsoner** [comment]
Hmm... I looked quickly and I think it's specific to ICP actually. It updates head even though it doesn't need to. I'll push a fix

Let me know if you're looking into the VTK stuff, happy to let you do it but also interested to see if it helps :)

**Comment by GuillaumeFavelier** [comment]
`fill_holes` is there too but I'm a bit worried of:

image

**Comment by GuillaumeFavelier** [comment]
My changes are not compatible with the changes of `clim`

**Comment by larsoner** [comment]
> My changes are not compatible with the changes of clim

If you want to force push over 80bcc775828a27f27818d55a9607bc7287064129 feel free, I can pull, cherry pick, and resolve

**Comment by GuillaumeFavelier** [comment]
I just reverted for now.

**Comment by GuillaumeFavelier** [comment]
It's too optimistic for now. I'll add a run of `fill_holes` of a size around 1/5th of the extent of the mesh.

**Comment by GuillaumeFavelier** [comment]
I ended up disabling `check_surface` because it is too strict.

**Comment by larsoner** [comment]
FYI I found a bug with the scale mode stuff. It looks like the values in the double spin boxes are not updated properly during ICP. For uniform mode no values are updated (they are all set to 100), for 3-axis mode, only Y is updated (X and Z are stuck at 100):


I have a fix for the recomputation of `_check_inside` in ICP mode. I don't think we need the property stuff... okay for me to push?

> It's too optimistic for now. I'll add a run of fill_holes of a size around 1/5th of the extent of the mesh.

FWIW in principle these meshes *should* be topologically correct. Any holes should be small and hopefully easily fixed. First can we check to see if `vtkSelectEnclosedPoints` would actually yield any speed gains on Windows? If it doesn't, we don't need to bother with any hole filling stuff. If it does help, then we can work on topological defects...

**Comment by larsoner** [comment]
Forgot to post the video of the bug:

https://user-images.githubusercontent.com/2365790/157269864-7ff78586-6c10-4f1e-b573-2e86f1e09ca1.mp4

**Comment by GuillaumeFavelier** [comment]
> we check to see if vtkSelectEnclosedPoints would actually yield any speed gains on Windows?

Sorry, I forgot to communicate on that and jumped on the other bugs 😅 

So yes, I can confirm that it's faster. I'll fill in a table like https://github.com/mne-tools/mne-python/pull/10377#issuecomment-1060833766

**Comment by GuillaumeFavelier** [comment]
> okay for me to push?

Go for it @larsoner, I'm on https://github.com/mne-tools/mne-python/pull/10377#issuecomment-1061905404

**Comment by larsoner** [comment]
Okay I pushed to unify. Since for the coreg we know we have PyVista, we should probably use `mode='pyvista'`. For other parts of the code like forward we probably shouldn't, since not all users will have VTK. I'll push a commit for that, too!

**Comment by GuillaumeFavelier** [comment]
Thank you for the nice integration

**Comment by larsoner** [comment]
... this new code is fast enough that we can actually use `_bem_high_res` for the inside check if we want. In fact it probably makes sense to make the scaling and inside/outside actually match which surface is being used/shown (high or low res). Should I push a commit for this, too?

**Comment by GuillaumeFavelier** [comment]
>  this new code is fast enough that we can actually use _bem_high_res for the inside check if we want

I suppose we can afford it now...

> Should I push a commit for this, too?

I can run another benchmark on Windows :+1:

**Comment by larsoner** [comment]
Okay @GuillaumeFavelier feel free to take a look and see if all the code changes make sense to you. If so, feel free to remove WIP, mark ready for review, and I'll merge once things are green!

File: mne/gui/_coreg.py [hunk 3]
```diff
@@ -1016,7 +1025,8 @@ def _set_sensors_visibility(self, state):
 
     def _update_actor(self, actor_name, actor):
         # XXX: internal plotter/renderer should not be exposed
-        self._renderer.plotter.remove_actor(self._actors.get(actor_name))
+        self._renderer.plotter.remove_actor(self._actors.get(actor_name),
+                                            render=False)
```
Comment (larsoner) [inline mne/gui/_coreg.py:1029]: FYI this made me realize we had a lot of these elsewhere. This gets rid of the bug where things "flash" a bit during updates. Now they just update as I'd expect!

PR to fix other instances is in https://github.com/mne-tools/mne-python/pull/10423

Comment (GuillaumeFavelier) [reply mne/gui/_coreg.py:1029]: > This gets rid of the bug where things "flash" a bit during updates.

I did not notice those but I'm glad it's taken care of

**Comment by GuillaumeFavelier** [comment]
Latest benchmark, after (4b244b47adfd5e93db271278db489e6a98cbfd63):

x | main | PR
--|--|--
linux | 2.36s | -
windows | 160.85s | 2.4s

Slightly slower on Windows but I think it's worth it considering what we gain with the more accurate calculation. Plus, what is timed is all the drawing at init and my CPU is an old dual core, I can live with that.
We can always optimize the 3d plots further.

**Comment by larsoner** [comment]
Argh I can replicate a failure locally, will fix and push

**Comment by larsoner** [comment]
Thanks @GuillaumeFavelier !

**Pull request merged**